src/main/java/org/jitsi/videobridge/BitrateController.java

```diff
@@ -0,0 +1,613 @@
+     */
+    boolean isForwarded(RtpChannel source)
+    {
+        String endpointID = source.getEndpoint().getID();
+        return forwardedEndpoints.contains(endpointID);
+    }
+
+    /**
+     * Computes the optimal and the target bitrate, limiting the target to be
+     * less than bandwidth estimation specified as an argument.
+     *
+     * @param bweBps the max bandwidth estimation that the target bitrate must
+     * not exceed.
+     *
+     * @return an array of {@link EndpointBitrateAllocation}.
+     */
+    private EndpointBitrateAllocation[] allocate(
+        long bweBps, List<Endpoint> conferenceEndpoints)
+    {
+        // Init.
+        int szConference = conferenceEndpoints.size();
+
+        // subtract 1 for destEndpoint.
+        EndpointBitrateAllocation[] endpointBitrateAllocations
+            = new EndpointBitrateAllocation[szConference - 1];
+
+        int lastN = dest.getLastN();
+        if (lastN < 0)
+        {
+            // If lastN is disable, pretend lastN == szConference.
+            lastN = endpointBitrateAllocations.length;
+        }
+
+        Endpoint destEndpoint = dest.getEndpoint();
+        Set<String> pinnedEndpoints = destEndpoint.getPinnedEndpoints();
+        int szPinned = pinnedEndpoints == null ? 0 : pinnedEndpoints.size();
+        if (szPinned > 0)
+        {
+            // The lastN value should be large enough to accommodate all the
+            // pinned endpoints.
+            lastN = Math.max(lastN, szPinned);
+        }
+
+        Set<String> selectedEndpoints = destEndpoint.getSelectedEndpoints();
+        int offPinned = 0, szSelected
+            = selectedEndpoints == null ? 0 : selectedEndpoints.size();
+
+        int maxQuality = 0;
+        long oldBweBps = bweBps;
+
+        int idx = -1;
+        for (int i = 0; i < szConference; i++)
+        {
+            Endpoint sourceEndpoint = conferenceEndpoints.get(i);
+            if (sourceEndpoint == destEndpoint)
+            {
+                continue;
+            }
+
+            idx++;
+
+            boolean pinned = szPinned != 0
+                && pinnedEndpoints.contains(sourceEndpoint.getID());
+
+            int priority; // Higher priority endpoints are optimized first.
+            if (pinned)
+            {
+                priority = offPinned;
+                offPinned++;
+            }
+            else
+            {
+                priority = idx - offPinned + szPinned;
+            }
+
+            boolean selected = szSelected != 0
+                && selectedEndpoints.contains(sourceEndpoint.getID());
+
+            boolean forwarded = priority < lastN;
+
+            EndpointBitrateAllocation endpointBitrateAllocation
+                = new EndpointBitrateAllocation(
+                sourceEndpoint, forwarded, selected);
+
+            endpointBitrateAllocations[priority] = endpointBitrateAllocation;
+
+            // First pass.
+            endpointBitrateAllocation.allocate(bweBps, maxQuality);
```

Comment:
What is the idea here? maxEndpoint gets incremented on each iteration, and we are not necessarily iterating according to priority.